feat!: Update to Groovy 4

matrei opened the pull request.

matrei committed:
feat!: Update to Groovy 4

matrei committed:
fix: Update to and compatible with Groovy 4.0.22

matrei committed:
fix: Update groovy-joint-workflow to use Groovy 4

matrei committed:
fix: Update groovy-joint-workflow

The GROOVY_4_0_X branch has split the Develocity config
into multiple files. This commit is meant to accommodate for that.

sdelamo reviewed (changes requested):
Could you change the failing test from @Ignore to  PendingFeature instead?

matrei committed:
test: Replace @Ignore with @PendingFeature where possible

One JUnit test is replaced with Spock to be able to use `@PendingFeature`
as I could not find an equivalent annotation in JUnit.

matrei committed:
test: Remove @Ignore on passing test

This test passes now, not sure why it failed before

matrei committed:
chore(test): Code cleanup

sdelamo commented on grails-test-suite-uber/src/test/groovy/grails/test/mixin/InheritanceWithValidationTests.groovy:
```diff
@@ -47,12 +49,12 @@ class AbstractCustomPropertyValue {
 }
 
-@Entity
+//@Entity
```
why this ?

matrei replied:
The test won't compile with the `@Entity` annotation present when extending another domain class.

We could also do this:
```groovy
// With Groovy 4, it is currently not possible to extend domain classes: https://issues.apache.org/jira/browse/GROOVY-5106
@Entity
class StringPropertyValue /*extends AbstractCustomPropertyValue*/ {

    String stringValue

    static constraints = {
        stringValue (nullable: true)
    }

    StringPropertyValue (String value) {
        this.stringValue = value
        this.valid = true
    }
}

```

sdelamo commented on grails-test-suite-uber/src/test/groovy/grails/validation/DomainConstraintGettersSpec.groovy:
```diff
@@ -67,6 +68,7 @@ class DomainConstraintGettersSpec extends Specification implements DataTest {
 
     // DOMAIN WITH SUPER CLASS
 
+    @PendingFeature(reason = 'With Groovy 4, it is currently not possible to extend domain classes: https://issues.apache.org/jira/browse/GROOVY-5106')
```
https://issues.apache.org/jira/browse/GROOVY-5106 seems to be fixed. are we not using a groovy version which does not contain the patch?

sdelamo replied:
@paulk-asert do you know about this?

matrei replied:
It is the fix to that issue that causes the problem with extending domain classes.

paulk-asert replied:
If you have the following classes in Java:

    class Parent implements Iterator<CharSequence> { ... }
    class Foo extends Parent implements Iterator<String> { ... }

Java will tell you to remove one of the Iterator interface clauses:

    'java.util.Iterator' cannot be inherited with different type arguments: 'java.lang.CharSequence' and 'java.lang.String'. 

Groovy 3 allowed that to co-exist since after type erasure, it folds down onto one interface and the one you want was kept. But that was a kludge and information was lost, so Groovy 4 does the same as Java here.

In this case we have (note duplicated `GormEntity` and `Entity` with different types):

    class AbstractCustomPropertyValue implements GroovyObject, GormEntity<AbstractCustomPropertyValue>, GormEntityDirtyCheckable, WebDataBinding, DomainClass, grails.gorm.Entity<AbstractCustomPropertyValue>, DirtyCheckable.Trait.FieldHelper, GormValidateable.Trait.FieldHelper { ... }
    class StringPropertyValue extends AbstractCustomPropertyValue implements GormEntity<StringPropertyValue>, WebDataBinding, DomainClass, grails.gorm.Entity<StringPropertyValue>, DirtyCheckable.Trait.FieldHelper, GormValidateable.Trait.FieldHelper { ... }

If supporting inheritance of entities is desirable, the `@Entity` transform would need to be made smarter.  There might be a few different ways to make it work. Perhaps only abstract classes could be extended and they might not need the original interface added. Perhaps non-final classes could use a generic type like `? extends AbstractCustomProperty` (in this case).

paulk-asert replied:
My example was for the previous test but the discussion still applies. For the previous test, I think you can also remove `@Entity` from the abstract parent class (and remove it from `mockDomains` too) and the test still passes prior to this PR at least.

matrei replied:
Yes, you are right, the test succeeds if `@Entity` is removed from `AbstractCustomPropertyValue` (and it also is removed from mockDomains).

sdelamo replied:
@matrei can we update the test and start a breaking changes section in the docs telling the user about this change?

matrei replied:
@sdelamo Yes, we could. Are you aware of any use cases where extended domain classes are also used concretely? That kind of usage would obviously not work anymore.

puneetbehl replied:
You could try running https://github.com/grails/grails-functional-tests project to see if anything is broken.

matrei replied:
> You could try running https://github.com/grails/grails-functional-tests project to see if anything is broken.

Hi @puneetbehl, is it possible to run the `grails-functional-tests` project on the this branch?

puneetbehl replied:
Yes, you would need to make certain changes to the `build.gradle` file of the project.

sdelamo reviewed (commented):
@matrei Thanks for doing this PR. Moreover, thanks for doing such a granular PR. It makes review really easy.

matrei committed:
test: Remove `@Entity` on extended domain classes

Since Groovy 4 it is currently no longer possible to extend domain classes.

matrei committed:
test: Replace `@Ignore` with `@PendingFeature`

matrei committed:
test: Clean up and improve code readability

sdelamo reviewed (approved):
I am going to merge this into 7.0.x branch. We can work on the @PendingFeature tests in separate PRs. Thanks for the work.
